pkg/dwarf/loclist/loclist.go

```diff
@@ -0,0 +1,96 @@
+package loclist
+
+import (
+	"encoding/binary"
+)
+
+// Reader parses and presents DWARF loclist information.
+type Reader struct {
+	data  []byte
+	cur   int
+	ptrSz int
+}
+
+// New returns an initialized loclist Reader.
+func New(data []byte, ptrSz int) *Reader {
+	return &Reader{data: data, ptrSz: ptrSz}
+}
+
+// Empty returns true if this reader has no data.
+func (rdr *Reader) Empty() bool {
+	return rdr.data == nil
+}
+
+// Seek moves the data pointer to the specified offset.
+func (rdr *Reader) Seek(off int) {
+	rdr.cur = off
+}
+
+// Next advances the reader to the next loclist entry, returning
+// the entry and true if successful, or nil, false if not.
+func (rdr *Reader) Next(e *Entry) bool {
+	e.lowpc = rdr.oneAddr()
+	e.highpc = rdr.oneAddr()
+
+	if e.lowpc == 0 && e.highpc == 0 {
+		return false
+	}
+
+	if e.BaseAddressSelection() {
+		e.instr = nil
+		return true
+	}
+
+	instrlen := binary.LittleEndian.Uint16(rdr.read(2))
+	e.instr = rdr.read(int(instrlen))
+	return true
+}
+
+func (rdr *Reader) read(sz int) []byte {
+	r := rdr.data[rdr.cur : rdr.cur+sz]
+	rdr.cur += sz
+	return r
+}
+
+func (rdr *Reader) oneAddr() uint64 {
+	switch rdr.ptrSz {
+	case 4:
+		addr := binary.LittleEndian.Uint32(rdr.read(rdr.ptrSz))
+		if addr == ^uint32(0) {
+			return ^uint64(0)
+		}
+		return uint64(addr)
+	case 8:
+		addr := uint64(binary.LittleEndian.Uint64(rdr.read(rdr.ptrSz)))
+		return addr
+	default:
+		panic("bad address size")
+	}
+}
+
+// Entry represents a single entry in the loclist section.
+type Entry struct {
+	lowpc, highpc uint64
```

Comment:
I think I prefer these and instr to be public and not have reader methods. Like debug/dwarf.LineEntry in the std library.